### PR TITLE
pyc: add arg reset-m-time

### PR DESCRIPTION
**Pull request by prestist**
In ostree systems, the mtime gets confused and results in a re-compile of the .pyc files due to how the files are linked. Add support to identify and set mtime to zero to avoid this issue.

### CODE REVIEW

**Comment by prestist** [comment]
@keszybz we are wanting to leverage add-determinism to rpm-ostree and ostree-rs-ext and were wondering if there would be a way to use the current version of add-determinism as a crate or if we needed to expand some of the calls by making it more suitable for this. 

The two use cases we see listed in the readme dont really follow ours. 

What are your thoughts? and if its okay can we expand the api to allow it to be consumed in this manner?

File: src/handlers/pyc.rs [hunk 1]
```diff
@@ -700,8 +700,28 @@ impl PycParser {
         assert_eq!(data == self.data, removed_count == 0);
         Ok((removed_count > 0, data))
     }
+    
+    fn clear_mtime(&mut self) -> Result<(bool, Vec<u8>)> {
```
Comment (keszybz) [inline src/handlers/pyc.rs:704]: Just pass the existing `data` and `have_mod` as refs here, no need to clone.

**Comment by keszybz** [comment]
> What are your thoughts? and if its okay can we expand the api to allow it to be consumed in this manner?

Yes.

> In ostree systems, the mtime gets confused and results in a re-compile of the .pyc files due to how the files are linked. Add support to identify and set mtime to zero to avoid this issue.

Can you explain what the issue is? I thought ostree uses mtime of 0, but I'd like to understand the issue more fully.

**Comment by prestist** [comment]
> > What are your thoughts? and if its okay can we expand the api to allow it to be consumed in this manner?
> 
> Yes.
> 
> > In ostree systems, the mtime gets confused and results in a re-compile of the .pyc files due to how the files are linked. Add support to identify and set mtime to zero to avoid this issue.
> 
> Can you explain what the issue is? I thought ostree uses mtime of 0, but I'd like to understand the issue more fully.

Sure, its a looong thread found here https://github.com/ostreedev/ostree/issues/1469

File: src/handlers/pyc.rs [hunk 2]
```diff
@@ -713,21 +733,29 @@ impl super::Processor for Pyc {
 
     fn process(&self, input_path: &Path) -> Result<super::ProcessResult> {
         let (mut io, input) = InputOutputHelper::open(input_path, self.config.check)?;
-
+
```
Comment (travier) [inline src/handlers/pyc.rs:736]: Nit: some whitespace error here

File: src/handlers/pyc.rs [hunk 3]
```diff
@@ -713,21 +733,29 @@ impl super::Processor for Pyc {
...
     fn process(&self, input_path: &Path) -> Result<super::ProcessResult> {
         let (mut io, input) = InputOutputHelper::open(input_path, self.config.check)?;
-
+        
         let mut parser = PycParser::from_file(input_path, input)?;
         if parser.version < (3, 0) {
             return Ok(super::ProcessResult::Noop);  // We don't want to touch python2 files
         }
-
+
```
Comment (travier) [inline src/handlers/pyc.rs:741]: idem

**Comment by travier** [comment]
Let's add the following links to this PR message/commit message to help cross-reference things:
- https://gitlab.com/fedora/bootc/tracker/-/issues/3
- https://github.com/ostreedev/ostree/issues/1469
- https://pagure.io/workstation-ostree-config/pull-request/505

**Comment by keszybz** [comment]
Please see #27. It implements this idea, but the implementation is quite different.

**Comment by prestist** [comment]
> Please see #27. It implements this idea, but the implementation is quite different.

Firstly, thank you for taking the time to implement it. It was fun to try and figure out how to work with this repo, sorry if we did not follow the form of it. Your changes look appropriate for our use. The only thing outstanding would be a way to consume the project via rust rather then cli.

**Comment by prestist** [comment]
Closing as it is being replaced by #27

**Pull request closed without merging**